### PR TITLE
Log provider improvements

### PR DESCRIPTION
**Pull request by armando-rodriguez-cko**
- Log provider improvements
- Add test to LogProvider

### CODE REVIEW

**Review by david-chandler-cko — COMMENTED** [review]
Since the logger factory already caches the logger instances, I think this class could be simplified right down to the following:

```csharp
public static class LogProvider
{
    private static ILoggerFactory _loggerFactory = new LoggerFactory();

    public static void SetLogFactory(ILoggerFactory factory)
    {
        _loggerFactory?.Dispose();
        _loggerFactory = factory;
    }

    public static ILogger GetLogger(Type loggerType)
    {
        return loggerType is null ? NullLogger.Instance : _loggerFactory.CreateLogger(loggerType);
    }
}
```

**Comment by IanKemp** [comment]
Any movement on this?

**Comment by armando-rodriguez-cko** [comment]
> Any movement on this?

I have this task in the current sprint, but if I don't have time, next sprint I will try to implement it.

**Comment by leandro-tylkovitch-cko** [comment]
> Since the logger factory already caches the logger instances, I think this class could be simplified right down to the following:
> 
> ```cs
> public static class LogProvider
> {
>     private static ILoggerFactory _loggerFactory = new LoggerFactory();
> 
>     public static void SetLogFactory(ILoggerFactory factory)
>     {
>         _loggerFactory?.Dispose();
>         _loggerFactory = factory;
>     }
> 
>     public static ILogger GetLogger(Type loggerType)
>     {
>         return loggerType is null ? NullLogger.Instance : _loggerFactory.CreateLogger(loggerType);
>     }
> }
> ```

David,

I tried your approach but Integration Testing is failing. I've added your proposed Unit Test for ShouldCreateASingleLoggerInstanceForMultipleConcurrentRequests and it's working OK with the actual propposal

**Comment by david-chandler-cko** [comment]
> > Since the logger factory already caches the logger instances, I think this class could be simplified right down to the following:
> > ```cs
> > public static class LogProvider
> > {
> >     private static ILoggerFactory _loggerFactory = new LoggerFactory();
> > 
> >     public static void SetLogFactory(ILoggerFactory factory)
> >     {
> >         _loggerFactory?.Dispose();
> >         _loggerFactory = factory;
> >     }
> > 
> >     public static ILogger GetLogger(Type loggerType)
> >     {
> >         return loggerType is null ? NullLogger.Instance : _loggerFactory.CreateLogger(loggerType);
> >     }
> > }
> > ```
> 
> David,
> 
> I tried your approach but Integration Testing is failing. I've added your proposed Unit Test for ShouldCreateASingleLoggerInstanceForMultipleConcurrentRequests and it's working OK with the actual propposal

Do you have an example of the failure when you simplify the code? What's the integration test trying to do?

File: src/CheckoutSdk/LogProvider.cs
Comment (david-chandler-cko) [inline src/CheckoutSdk/LogProvider.cs:21]:
```suggestion
            return loggerType is null ? NullLogger.Instance : _loggerFactory.CreateLogger(loggerType) ?? NullLogger.Instance;
```

File: test/CheckoutSdkTest/Extensions/CheckoutServiceCollectionTest.cs
Comment (david-chandler-cko) [inline test/CheckoutSdkTest/Extensions/CheckoutServiceCollectionTest.cs:19]: These test changes should be able to be reverted if you make the suggested change, as discussed on Slack

**Comment by david-chandler-cko** [comment]
I can't approve this as it's the code I suggested.

File: test/CheckoutSdkTest/Extensions/CheckoutServiceCollectionTest.cs
Comment (armando-rodriguez-cko) [reply test/CheckoutSdkTest/Extensions/CheckoutServiceCollectionTest.cs]: 👍🏼

File: src/CheckoutSdk/LogProvider.cs
Comment (armando-rodriguez-cko) [reply src/CheckoutSdk/LogProvider.cs]: Thank you @david-chandler-cko

File: test/CheckoutSdkTest/Extensions/CheckoutServiceCollectionTest.cs
Comment (armando-rodriguez-cko) [reply test/CheckoutSdkTest/Extensions/CheckoutServiceCollectionTest.cs]: 👍🏼

**Comment by david-chandler-cko** [comment]
A couple of points:
1. The suggested test got removed: https://github.com/checkout/checkout-sdk-net/pull/375#issuecomment-1980865741
2. The code seems fine, but I don't think I should approve this PR as it's just the code that I suggested should be pasted in.

**Review by diego-apestegui-cko — APPROVED** [review]

**Comment by armando-rodriguez-cko** [comment]
> A couple of points:
> 
> 1. The suggested test got removed: [Log provider improvements #375 (comment)](https://github.com/checkout/checkout-sdk-net/pull/375#issuecomment-1980865741)
> 2. The code seems fine, but I don't think I should approve this PR as it's just the code that I suggested should be pasted in.

Yes I forgot this file, sorry

**Review by leandro-tylkovitch-cko — APPROVED** [review]

**Pull request merged**
